src/resources/ble_darwin.mm

```diff
@@ -0,0 +1,1297 @@
+    : BLEResource(group, SERVICE)
+    , _service([service retain])
+    , _device(device)
+    , _peripheral_manager(null)
+    , _characteristics_resource_index([[NSMutableDictionary new] retain])
+    , _deployed(false) {}
+
+  BLEServiceResource(BLEResourceGroup* group, BLEPeripheralManagerResource* peripheral_manager, CBService* service)
+      : BLEResource(group, SERVICE)
+      , _service([service retain])
+      , _device(null)
+      , _peripheral_manager(peripheral_manager)
+      , _characteristics_resource_index([[NSMutableDictionary new] retain])
+      , _deployed(false) {}
+
+  ~BLEServiceResource() override {
+    [_service release];
+    [_characteristics_resource_index release];
+  }
+
+  void make_deletable() override;
+
+  CBService* service() { return _service; }
+  BLERemoteDeviceResource* device() { return _device; }
+  BLEPeripheralManagerResource* peripheral_manager() { return _peripheral_manager; }
+  BLECharacteristicResource* get_or_create_characteristic_resource(CBCharacteristic* characteristic, bool can_create= false);
+  bool deployed() const { return _deployed; }
+  void set_deployed(bool deployed) { _deployed = deployed; }
+ private:
+  CBService* _service;
+  BLERemoteDeviceResource* _device;
+  BLEPeripheralManagerResource* _peripheral_manager;
+  NSMutableDictionary<CBUUID*, BLEResourceHolder*>* _characteristics_resource_index;
+  bool _deployed;
+};
+
+class CharacteristicData;
+typedef DoubleLinkedList<CharacteristicData> CharacteristicDataList;
+class CharacteristicData: public CharacteristicDataList::Element {
+ public:
+  explicit CharacteristicData(NSData *data): _data([data retain]) {}
+  ~CharacteristicData() { [_data release]; }
+  NSData* data() { return _data; }
+ private:
+  NSData* _data;
+};
+
+class BLECharacteristicResource : public BLEResource, public DiscoverableResource {
+ public:
+  TAG(BLECharacteristicResource);
+
+  BLECharacteristicResource(BLEResourceGroup* group, BLEServiceResource *service, CBCharacteristic* characteristic)
+      : BLEResource(group, CHARACTERISTIC)
+      , _characteristic([characteristic retain])
+      , _characteristic_data_list()
+      , _service(service)
+      , _subscriptions([[NSMutableArray new] retain]) {}
+
+  ~BLECharacteristicResource() override {
+    [_characteristic release];
+    [_subscriptions release];
+  }
+
+  CBCharacteristic* characteristic() { return _characteristic; }
+
+  void set_error(NSError* error) {
+    if (_error != nil) [_error release];
+    if (error != nil) error = [error retain];
+    _error = error;
+  }
+
+  NSError* error() {
+    return _error;
+  }
+
+  void append_data(NSData* data) {
+    _characteristic_data_list.append(_new CharacteristicData(data));
+  }
+
+  CharacteristicData *remove_first() {
+    return _characteristic_data_list.remove_first();
+  }
+
+  void put_back(CharacteristicData* data) {
+    _characteristic_data_list.prepend(data);
```

Comment:
Feels weird that the function is called `put_back` but we are calling `prepend` here.

Reply:
suggestion?

Reply:
Maybe `unget`.
But Erik interpreted the name `put_back` as just returning it. I read it similar to `push_back` in C++...